### PR TITLE
Working to fix nodeos_snapshot_diff_test by sync on trx gen startup

### PR DESCRIPTION
**Pull request by oschwaldp-oci**
Refactor - move `launchTrxGenerators` from `Node.py` to `Cluster.py` where it fits better.

Add `waitForTrxGeneratorsSpinup` function which allows tests to wait for initial trx from each generator to appear in a block.  This allows the tests to sync on the trx generators spinning up and knowing that trxs are flowing before proceeding with tests.

Update tests that used `launchTrxGenerators` for the refactor and to use the `waitForTrxGeneratorsSpinup` function.

`trx_generator` now logs first transaction it is preparing to send right before sending it to allow synchronization by tests without influencing its generation rate since it occurs prior only on initial trx and prior to initial trx send.

Resolves: #637

### CODE REVIEW

File: tests/nodeos_contrl_c_test.py [hunk 1]
```diff
@@ -98,16 +98,16 @@
...
-                                    contractOwnerAcctName=cluster.eosioAccount.name, acctNamesList=[accounts[0].name,accounts[1].name],
-                                    acctPrivKeysList=[account1PrivKey,account2PrivKey], p2pListenPort=nonProducerP2pPort, waitToComplete=False)
+    cluster.launchTrxGenerators(contractOwnerAcctName=cluster.eosioAccount.name, acctNamesList=[accounts[0].name,accounts[1].name],
+                                acctPrivKeysList=[account1PrivKey,account2PrivKey], nodeId=nonProdNode.nodeId, tpsPerGenerator=targetTpsPerGenerator,
+                                numGenerators=trxGeneratorCnt, durationSec=testTrxGenDurationSec, waitToComplete=False)
 
     Print("Give txn generator time to spin up and begin producing trxs")
-    time.sleep(10)
+    cluster.waitForTrxGeneratorsSpinup(nodeId=nonProdNode.nodeId, numGenerators=trxGeneratorCnt)
+    time.sleep(5)
```
Comment (heifner) [inline tests/nodeos_contrl_c_test.py:110]: This sleep is no longer needed, right?

File: tests/nodeos_snapshot_diff_test.py [hunk 2]
```diff
@@ -120,12 +120,13 @@ def waitForBlock(node, blockNum, blockType=BlockType.head, timeout=None, reportI
     waitForBlock(node0, blockNum, blockType=BlockType.lib)
 
     Print("Configure and launch txn generators")
-    producerP2pPort = cluster.getNodeP2pPort(snapshotNodeId)
     targetTpsPerGenerator = 667
```
Comment (heifner) [inline tests/nodeos_snapshot_diff_test.py:123]: This is way overkill, 10 should be more than enough. Just need a few.

File: tests/nodeos_contrl_c_test.py [hunk 3]
```diff
@@ -98,16 +98,16 @@
     testSuccessful=False
 
     Print("Configure and launch txn generators")
-    nonProducerP2pPort = cluster.getNodeP2pPort(nonProdNode.nodeId)
     targetTpsPerGenerator = 500
```
Comment (heifner) [inline tests/nodeos_contrl_c_test.py:101]: We can reduce this. Maybe 100 would be better.

File: tests/trx_generator/trx_generator.cpp [hunk 4]
```diff
@@ -97,8 +97,21 @@ namespace eosio::testing {
...
+      std::ofstream out(fileName.str());
+
+      out << fc::string(trx.id()) << "\n";
+      out.close();
+   }
+
    void transfer_trx_generator::push_transaction(p2p_trx_provider& provider, signed_transaction_w_signer& trx, uint64_t& nonce_prefix, uint64_t& nonce, const fc::microseconds& trx_expiration, const chain_id_type& chain_id, const block_id_type& last_irr_block_id) {
       update_resign_transaction(trx._trx, trx._signer, ++nonce_prefix, nonce, trx_expiration, chain_id, last_irr_block_id);
+      if (_txcount == 0)
+      {
```
Comment (heifner) [inline tests/trx_generator/trx_generator.cpp:112]: Odd, change of style for this `if`.

**Review by heifner — CHANGES_REQUESTED** [review]

File: tests/nodeos_snapshot_diff_test.py (hunk 2 above)
Comment (oschwaldp-oci) [reply tests/nodeos_snapshot_diff_test.py:123]: Happy to reduce it. Original aim was to maintain generation rates original attempted when using the trx_test_gen_plugin.

File: tests/nodeos_contrl_c_test.py (hunk 3 above)
Comment (oschwaldp-oci) [reply tests/nodeos_contrl_c_test.py:101]: Happy to reduce it. Original aim was to maintain generation rates original attempted when using the trx_test_gen_plugin.

File: tests/nodeos_snapshot_diff_test.py (hunk 2 above)
Comment (heifner) [reply tests/nodeos_snapshot_diff_test.py:123]: I created this test, no idea why I had it that large. Probably a copy/paste from somewhere else. Only need a couple of trx per block.

File: tests/nodeos_contrl_c_test.py (hunk 3 above)
Comment (heifner) [reply tests/nodeos_contrl_c_test.py]: Understood. Might as well reduce the load of this test while you are in here. The idea of the test is to have some transactions running during shutdown. 50 per block seems plenty for that requirement.

File: tests/trx_generator/trx_generator.cpp (hunk 4 above)
Comment (oschwaldp-oci) [reply tests/trx_generator/trx_generator.cpp]: Addressed: [de3a1c8](https://github.com/AntelopeIO/leap/pull/639/commits/de3a1c8392564d6d0ab41412374948c4c8f5359c)

File: tests/nodeos_snapshot_diff_test.py (hunk 2 above)
Comment (oschwaldp-oci) [reply tests/nodeos_snapshot_diff_test.py]: Addressed: [de3a1c8](https://github.com/AntelopeIO/leap/pull/639/commits/de3a1c8392564d6d0ab41412374948c4c8f5359c)

File: tests/nodeos_contrl_c_test.py (hunk 1 above)
Comment (oschwaldp-oci) [reply tests/nodeos_contrl_c_test.py]: Addressed: [de3a1c8](https://github.com/AntelopeIO/leap/pull/639/commits/de3a1c8392564d6d0ab41412374948c4c8f5359c)

Comment (heifner) [reply tests/nodeos_contrl_c_test.py]: It was just waiting before to make sure trx were flowing. So I think you can remove the sleep with your better wait for first trx.

**Review by heifner — APPROVED** [review]

**Comment by heifner** [comment]
Test failure:
```
2023-01-16T16:13:16.774983  Found first trx record: var/lib/first_trx_1338.txt
2023-01-16T16:13:16.775182  first transactions: ['7297b739c41a14c6ba115913d5bd63b0e690825060450b8e31e0ff6f81573756']
2023-01-16T16:14:14.449017             FAILURE - Exception during "get transaction_trace". Exception message: Error 3200005: http request fail
Error Details:
Error code 404
: {"code":404,"message":"Trace API: transaction id missing in the transaction id log files","error":{"code":0,"name":"","what":"","details":[]}}
```

File: tests/nodeos_contrl_c_test.py (hunk 1 above)
Comment (oschwaldp-oci) [reply tests/nodeos_contrl_c_test.py]: Actually can't do the wait here due to all production nodes being shut down.

**Review by heifner — APPROVED** [review]

**Review by linh2931 — APPROVED** [review]

**Pull request merged**